lucet-runtime/lucet-runtime-internals/src/c_api.rs

```diff
@@ -218,117 +224,113 @@ pub struct CYieldedVal {
                                 reason: lucet_terminated_reason::CtxNotFound,
-                                provided: std::ptr::null_mut(),
+                                provided: ptr::null_mut(),
                             },
                             TerminationDetails::YieldTypeMismatch => lucet_terminated {
                                 reason: lucet_terminated_reason::YieldTypeMismatch,
-                                provided: std::ptr::null_mut(),
+                                provided: ptr::null_mut(),
                             },
                             TerminationDetails::BorrowError(_) => lucet_terminated {
                                 reason: lucet_terminated_reason::BorrowError,
-                                provided: std::ptr::null_mut(),
+                                provided: ptr::null_mut(),
                             },
                             TerminationDetails::Provided(p) => lucet_terminated {
                                 reason: lucet_terminated_reason::Provided,
                                 provided: p
                                     .downcast_ref()
                                     .map(|CTerminationDetails { details }| *details)
-                                    .unwrap_or(std::ptr::null_mut()),
+                                    .unwrap_or(ptr::null_mut()),
                             },
                         },
                     },
                 },
-                State::Yielded { val, .. } => lucet_state {
-                    tag: lucet_state_tag::Yielded,
-                    val: lucet_state_val {
-                        yielded: lucet_yielded {
-                            val: val
-                                .downcast_ref()
-                                .map(|CYieldedVal { val }| *val)
-                                .unwrap_or(std::ptr::null_mut()),
-                        },
-                    },
+                Err(e) => lucet_result {
+                    tag: lucet_result_tag::Errored,
+                    val: lucet_result_val { errored: e.into() },
                 },
             }
         }
     }
 
     #[repr(C)]
     #[derive(Clone, Copy)]
-    pub struct lucet_state {
-        pub tag: lucet_state_tag,
-        pub val: lucet_state_val,
+    pub struct lucet_result {
+        pub tag: lucet_result_tag,
+        pub val: lucet_result_val,
     }
 
     #[repr(C)]
     #[derive(Clone, Copy, Debug, FromPrimitive)]
-    pub enum lucet_state_tag {
+    pub enum lucet_result_tag {
         Returned,
-        Running,
-        Fault,
-        Terminated,
         Yielded,
+        Faulted,
+        Terminated,
+        Errored,
     }
 
     #[repr(C)]
     #[derive(Clone, Copy)]
-    pub union lucet_state_val {
+    pub union lucet_result_val {
         pub returned: lucet_val::lucet_untyped_retval,
-        // no meaning to this boolean, it's just there so the type is FFI-safe
```

Comment:
What was the deal with this, I'd have thought that the union would be FFI-safe by virtue of `#[repr(C)]`?

Reply:
"safe" is maybe the wrong word; this is because C doesn't allow for unions with empty variants